File: articles/co-design-participatory-design/go/participation_quality_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type CodesignActivity struct {
	Name                         string
	Representation               float64
	Accessibility                float64
	ParticipantInfluence         float64
	TrustQuality                 float64
	EvidenceQuality              float64
	ImplementationAccountability float64
	DecisionImpact               float64
	EthicalRisk                  float64
	ParticipationQuality         float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func participationQuality(x CodesignActivity) float64 {
	return 0.18*x.Representation +
		0.14*x.Accessibility +
		0.22*x.ParticipantInfluence +
		0.12*x.TrustQuality +
		0.12*x.EvidenceQuality +
		0.12*x.ImplementationAccountability +
		0.14*x.DecisionImpact -
		0.08*x.EthicalRisk
}

func loadActivities(path string) []CodesignActivity {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var items []CodesignActivity

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 11 {
			log.Fatalf("malformed row: %v", row)
		}

		x := CodesignActivity{
			Name:                         row[0],
			Representation:               parseFloat(row[3]),
			Accessibility:                parseFloat(row[4]),
			ParticipantInfluence:         parseFloat(row[5]),
			TrustQuality:                 parseFloat(row[6]),
			EvidenceQuality:              parseFloat(row[7]),
			ImplementationAccountability: parseFloat(row[8]),
			DecisionImpact:               parseFloat(row[9]),
			EthicalRisk:                  parseFloat(row[10]),
		}
		x.ParticipationQuality = participationQuality(x)
		items = append(items, x)
	}

	return items
}

func main() {
	path := "../data/raw/codesign_activities_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	items := loadActivities(path)

	sort.Slice(items, func(i, j int) bool {
		return items[i].ParticipationQuality > items[j].ParticipationQuality
	})

	fmt.Println("rank,activity,participation_quality,ethical_risk")
	for i, x := range items {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, x.Name, x.ParticipationQuality, x.EthicalRisk)
	}
}

File: articles/co-design-participatory-design/cpp/participation_quality_engine.cpp
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

struct CodesignActivity {
    std::string name;
    double representation;
    double accessibility;
    double participant_influence;
    double trust_quality;
    double evidence_quality;
    double implementation_accountability;
    double decision_impact;
    double ethical_risk;
};

struct ScoredActivity {
    std::string name;
    double quality;
    double ethical_risk;
};

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> fields;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) {
        fields.push_back(item);
    }
    return fields;
}

double to_double(const std::string& value) {
    try {
        return std::stod(value);
    } catch (...) {
        throw std::runtime_error("Unable to parse numeric value: " + value);
    }
}

double participation_quality(const CodesignActivity& x) {
    return 0.18 * x.representation +
           0.14 * x.accessibility +
           0.22 * x.participant_influence +
           0.12 * x.trust_quality +
           0.12 * x.evidence_quality +
           0.12 * x.implementation_accountability +
           0.14 * x.decision_impact -
           0.08 * x.ethical_risk;
}

std::vector<CodesignActivity> read_activities(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Unable to open input file: " + path);
    }

    std::vector<CodesignActivity> activities;
    std::string line;
    std::getline(file, line);

    while (std::getline(file, line)) {
        if (line.empty()) {
            continue;
        }

        auto fields = split_csv_line(line);
        if (fields.size() < 11) {
            throw std::runtime_error("Malformed CSV row: " + line);
        }

        activities.push_back(
            CodesignActivity{
                fields[0],
                to_double(fields[3]),
                to_double(fields[4]),
                to_double(fields[5]),
                to_double(fields[6]),
                to_double(fields[7]),
                to_double(fields[8]),
                to_double(fields[9]),
                to_double(fields[10])
            }
        );
    }

    return activities;
}

int main(int argc, char* argv[]) {
    const std::string default_path = "../data/raw/codesign_activities_raw.csv";
    const std::string input_path = argc > 1 ? argv[1] : default_path;

    try {
        auto activities = read_activities(input_path);
        std::vector<ScoredActivity> scored;

        for (const auto& activity : activities) {
            scored.push_back(
                ScoredActivity{
                    activity.name,
                    participation_quality(activity),
                    activity.ethical_risk
                }
            );
        }

        std::sort(
            scored.begin(),
            scored.end(),
            [](const ScoredActivity& a, const ScoredActivity& b) {
                return a.quality > b.quality;
            }
        );

        std::cout << "rank,activity,participation_quality,ethical_risk\n";
        for (std::size_t i = 0; i < scored.size(); ++i) {
            std::cout << (i + 1) << ","
                      << scored[i].name << ","
                      << std::fixed << std::setprecision(6)
                      << scored[i].quality << ","
                      << scored[i].ethical_risk << "\n";
        }
    } catch (const std::exception& ex) {
        std::cerr << "ERROR: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}
